File: android/src/main/java/com/capgo/printer/PrinterPlugin.java
package com.capgo.printer;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "Printer")
public class PrinterPlugin extends Plugin {

    private final String pluginVersion = "8.0.13";

    private Printer implementation;

    @Override
    public void load() {
        implementation = new Printer(getContext(), getActivity());
    }

    @PluginMethod
    public void printBase64(PluginCall call) {
        String data = call.getString("data");
        String mimeType = call.getString("mimeType");
        String name = call.getString("name", "Document");

        if (data == null) {
            call.reject("data is required");
            return;
        }

        if (mimeType == null) {
            call.reject("mimeType is required");
            return;
        }

        try {
            implementation.printBase64(data, mimeType, name);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to print base64 data: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void printFile(PluginCall call) {
        String path = call.getString("path");
        String mimeType = call.getString("mimeType");
        String name = call.getString("name", "Document");

        if (path == null) {
            call.reject("path is required");
            return;
        }

        try {
            implementation.printFile(path, mimeType, name);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to print file: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void printHtml(PluginCall call) {
        String html = call.getString("html");
        String name = call.getString("name", "Document");

        if (html == null) {
            call.reject("html is required");
            return;
        }

        try {
            implementation.printHtml(html, name);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to print HTML: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void printPdf(PluginCall call) {
        String path = call.getString("path");
        String name = call.getString("name", "Document");

        if (path == null) {
            call.reject("path is required");
            return;
        }

        try {
            implementation.printPdf(path, name);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to print PDF: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void printWebView(PluginCall call) {
        String name = call.getString("name", "Document");

        try {
            implementation.printWebView(getBridge().getWebView(), name);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to print web view: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void getPluginVersion(final PluginCall call) {
        try {
            final JSObject ret = new JSObject();
            ret.put("version", this.pluginVersion);
            call.resolve(ret);
        } catch (final Exception e) {
            call.reject("Could not get plugin version", e);
        }
    }
}

File: ios/Sources/PrinterPlugin/PrinterPlugin.swift
import Foundation
import Capacitor

@objc(PrinterPlugin)
public class PrinterPlugin: CAPPlugin, CAPBridgedPlugin {
    private let pluginVersion: String = "8.0.13"
    public let identifier = "PrinterPlugin"
    public let jsName = "Printer"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "printBase64", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "printFile", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "printHtml", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "printPdf", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "printWebView", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = Printer()

    @objc func printBase64(_ call: CAPPluginCall) {
        guard let data = call.getString("data") else {
            call.reject("data is required")
            return
        }

        guard let mimeType = call.getString("mimeType") else {
            call.reject("mimeType is required")
            return
        }

        let name = call.getString("name") ?? "Document"

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            do {
                try self.implementation.printBase64(
                    data: data,
                    mimeType: mimeType,
                    name: name,
                    presentingViewController: self.bridge?.viewController
                )
                call.resolve()
            } catch {
                call.reject("Failed to print base64 data: \(error.localizedDescription)")
            }
        }
    }

    @objc func printFile(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject("path is required")
            return
        }

        let name = call.getString("name") ?? "Document"

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            do {
                try self.implementation.printFile(
                    path: path,
                    name: name,
                    presentingViewController: self.bridge?.viewController
                )
                call.resolve()
            } catch {
                call.reject("Failed to print file: \(error.localizedDescription)")
            }
        }
    }

    @objc func printHtml(_ call: CAPPluginCall) {
        guard let html = call.getString("html") else {
            call.reject("html is required")
            return
        }

        let name = call.getString("name") ?? "Document"

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            do {
                try self.implementation.printHtml(
                    html: html,
                    name: name,
                    presentingViewController: self.bridge?.viewController
                )
                call.resolve()
            } catch {
                call.reject("Failed to print HTML: \(error.localizedDescription)")
            }
        }
    }

    @objc func printPdf(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject("path is required")
            return
        }

        let name = call.getString("name") ?? "Document"

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }

            do {
                try self.implementation.printPdf(
                    path: path,
                    name: name,
                    presentingViewController: self.bridge?.viewController
                )
                call.resolve()
            } catch {
                call.reject("Failed to print PDF: \(error.localizedDescription)")
            }
        }
    }

    @objc func printWebView(_ call: CAPPluginCall) {
        let name = call.getString("name") ?? "Document"

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            guard let webView = self.bridge?.webView else {
                call.reject("WebView not available")
                return
            }

            do {
                try self.implementation.printWebView(
                    webView: webView,
                    name: name,
                    presentingViewController: self.bridge?.viewController
                )
                call.resolve()
            } catch {
                call.reject("Failed to print web view: \(error.localizedDescription)")
            }
        }
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve(["version": self.pluginVersion])
    }
}
